### PR TITLE
chore: Sync Endo

### PR DESCRIPTION
**Pull request by kriskowal**
This change brings Agoric SDK in sync with the latest versions from Endo as of December 12, 2023.

This includes syncing up breaking changes with interface guards. Where invalid call guards would silently allow an invocation previously, they now throw. Several interface guards needed fixing.

Some accommodations were necessary for the new `__getMethodNames__` and `__getInterfaceGuard__` protocols.

The algorithm for performing a speculative integration with an Endo branch required an adjustment to ensure that types were generated correctly.

For reasons opaque to the integrator, this required:
- many type imports to qualify the `.js` extension. The only relevant change from the Endo side is that many packages provide generated `.d.ts` types now.
- @agoric/time` to have a `types` field pointing at `index.js`, which by dint of fortunate coïncidence is a valid TypeScript definition file.
- a type error to be suppressed in `@agoric/zone` that only causes a type error in integration with `@agoric/vats`.

### CODE REVIEW

File: packages/zoe/test/unitTests/test-zoe.js
Comment (turadg) [inline packages/zoe/test/unitTests/test-zoe.js]: how will we know when this is no longer needed? is there a ticket? it should point here or this comment should point to it.

File: packages/inter-protocol/src/price/roundsManager.js
Comment (turadg) [inline packages/inter-protocol/src/price/roundsManager.js:104]: This should take one quote,
https://github.com/Agoric/agoric-sdk/blob/a271af98a48928e151404e622b64f4ce808e86f0/packages/zoe/tools/manualPriceAuthority.js#L39-L45

I'll push accurate guards.

File: packages/SwingSet/src/controller/controller.js
Comment (turadg) [inline packages/SwingSet/src/controller/controller.js:1]: > refactor: Make all TS import extensions explicit

Why?

File: packages/inter-protocol/src/price/fluxAggregatorKit.js
Comment (turadg) [inline packages/inter-protocol/src/price/fluxAggregatorKit.js]: this TODO is odd. perhaps it's meant to be a deprecated type?

File: packages/zone/src/durable.js
Comment (turadg) [inline packages/zone/src/durable.js:92]: I'll take a look at fixing this before approving

File: package.json
Comment (turadg) [inline package.json]: these 1.0s, will they all be semver? Are Endo maintainers committed to major very bump for all breaking changes? do deep imports count as the package API?

**Comment by erights** [comment]
> This change brings Agoric SDK in sync with the latest versions from Endo as of December 12, 2003.

Wow, that's a long time ago!

File: packages/SwingSet/src/controller/controller.js
Comment (kriskowal) [reply packages/SwingSet/src/controller/controller.js:1]: `Tsc` made me do it. I do not know why `tsc` insisted, but it did.

File: package.json
Comment (kriskowal) [reply package.json]: Deep imports are expressly forbidden by an `"exports"` directive in every public Endo package except `captp`, `far`, `lockdown`, and `marshal`, and these must be fixed.

Endo has for some time committed to backward-compatibility for dependents using the default carat (`^`) range, which is more strict than semver.

Comment (kriskowal) [reply package.json]: https://github.com/endojs/endo/pull/1907

Comment (kriskowal) [reply package.json]: Tracking the broader issue for Agoric SDK https://github.com/Agoric/agoric-sdk/issues/8679

File: packages/zoe/test/unitTests/test-zoe.js
Comment (kriskowal) [reply packages/zoe/test/unitTests/test-zoe.js]: @erights Does the need for the disjunct vanish due to the sync? I suspect we can get rid of it in this PR since syncing Endo eliminates the prior behavior, at least for the purposes of tests.

Comment (erights) [reply packages/zoe/test/unitTests/test-zoe.js]: Yes.

File: packages/inter-protocol/src/price/roundsManager.js
Comment (erights) [reply packages/inter-protocol/src/price/roundsManager.js:104]: Better guards would certainly be nice, but can happen after this PR. Thus, I won't withhold approval waiting for this conversation to be resolved.

File: packages/zone/src/durable.js
Comment (erights) [reply packages/zone/src/durable.js:92]: Likewise, although it would be nice to get this resolved now, it can happen after this PR, so I won't withhold approval waiting for it.

**Review by erights — APPROVED** [review]
LGTM!!!

File: packages/zoe/test/unitTests/test-zoe.js
Comment (kriskowal) [reply packages/zoe/test/unitTests/test-zoe.js]: Done.

**Pull request merged**
